src/librustc_parse/parser/item.rs

```diff
@@ -814,10 +813,9 @@ impl<'a> Parser<'a> {
 
     fn parse_ident_or_underscore(&mut self) -> PResult<'a, ast::Ident> {
         match self.token.kind {
-            token::Ident(name, false) if name == kw::Underscore => {
-                let span = self.token.span;
+            token::Ident(name @ kw::Underscore, false) => {
                 self.bump();
-                Ok(Ident::new(name, span))
+                Ok(Ident::new(name, self.prev_span))
```

Comment:
This is incorrect in presence of `$ident: ident` matchers, note how we do this "save, bump, create ident" dance in all the identifier parsing functions.

The reason is that `self.prev_span` is not the same thing as `self.token.span` of the previous token, see the comments on these fields.
Use of `self.prev_span` is actually subtly incorrect most cases, but for non-identifiers it only affects diagnostics, so it's used anyway. In case of identifiers it affects hygiene, so we have to be precise.

Reply:
Ah, thanks -- is there a regression test I could add to catch this mistake in the future? I'll add a comment as well.

Reply:
IIRC, there are regression tests for this for identifiers parsed in other contexts, but not for "parse ident or underscore".